### PR TITLE
Added the ConstAddressDeployer.

### PR DESCRIPTION
**Pull request by Foivos**

### CODE REVIEW

**Review by re1ro — CHANGES_REQUESTED** [review]
I think we need some tests for this one. Also how are we passing the constructor params to the deployed contract? I think we need that for our proxies

**Comment by Foivos** [comment]
> Also how are we passing the constructor params to the deployed contract?

Here is some js that uses ethers to deploy this contract and use it to deploy a new contract with some args. 

```js
const deployer = await deployContract(ownerWallet, ConstAddressDeployer);
const salt = keccak256(defaultAbiCoder.encode(['string'], [key]));
const factory = new ContractFactory(
  contractJSON.abi,
  contractJSON.bytecode
);
const bytecode = (await factory.getDeployTransaction(...args)).data;
await (await deployer.connect(wallet).deploy(bytecode, salt)).wait();
const address = await deployer.predict(bytecode, salt);
const contract = new Contract(address, contractJSON.abi, wallet);

```

There is however the following issue: If a contract needs to know params that vary between chains (like the gateway address on each chain) then those cannot be a part of the constructor, because that would cause a different bytecode and a different address as a result. For this reason contract wishing to have the same address. need to specify an empty constructor and an `init` function which will only be able to be called once to setup those variables after deployment.

File: src/ConstAddressDeployer.sol [hunk 1]
```diff
@@ -0,0 +1,23 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.9;
+
+contract ConstAddressDeployer {
+    event Deployed(bytes32 bytecodeHash, bytes32 salt, address deployedAddress);
+    function deploy(bytes memory bytecode, bytes32 salt) external returns (address deployedAddress) {
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            deployedAddress := create2(0, add(bytecode, 32), mload(bytecode), salt)
```
Comment (sammy1991106) [inline src/ConstAddressDeployer.sol:10]: Just to confirm, `add(bytecode, 32)` to skip its length?

File: src/ConstAddressDeployer.sol [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        }
+        emit Deployed(keccak256(bytecode), salt, deployedAddress);
+    }
+
+    function predict(bytes memory bytecode, bytes32 salt) external view returns (address deployedAddress) {
```
Comment (sammy1991106) [inline src/ConstAddressDeployer.sol:15]:
```suggestion
    function deployedAddress(bytes calldata bytecode, bytes32 salt) external view returns (address deployedAddress) {
```

File: src/ConstAddressDeployer.sol [hunk 3]
```diff
@@ -0,0 +1,23 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.9;
+
+contract ConstAddressDeployer {
+    event Deployed(bytes32 bytecodeHash, bytes32 salt, address deployedAddress);
+    function deploy(bytes memory bytecode, bytes32 salt) external returns (address deployedAddress) {
```
Comment (sammy1991106) [inline src/ConstAddressDeployer.sol:7]: Can bytecode reside in calldata here or it won't work with mload below?

Comment (sammy1991106) [inline src/ConstAddressDeployer.sol:7]:
```suggestion
    event Deployed(bytes32 bytecodeHash, bytes32 salt, address deployedAddress);
    
    function deploy(bytes memory bytecode, bytes32 salt) external returns (address deployedAddress) {
```

Comment (sammy1991106) [reply src/ConstAddressDeployer.sol:7]: Add an empty line between events and functions.

File: test/ConstAddressDeployer.js [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+      key, 
+      [name, symbol, decimals, capacity],
+    );
+    expect(await contract.address).to.equal(address);
+    expect(await contract.name()).to.equal(name);
+    expect(await contract.symbol()).to.equal(symbol);
+    expect(await contract.decimals()).to.equal(decimals);
+    expect(await contract.cap()).to.equal(capacity);
+  });
+  it('should deploy to the predicted address even with a different nonce', async () => {
```
Comment (sammy1991106) [inline test/ConstAddressDeployer.js:80]: Add an empty line between tests.

File: test/ConstAddressDeployer.js [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        to: userWallet.address,
+        value: 0,
+    });
+    expect(await contract.address).to.equal(address);
+    expect(await contract.name()).to.equal(name);
+    expect(await contract.symbol()).to.equal(symbol);
+    expect(await contract.decimals()).to.equal(decimals);
+    expect(await contract.cap()).to.equal(capacity);
+  });
+  it('should deploy the same contract twice to different addresses with different salts', async () => {
```
Comment (sammy1991106) [inline test/ConstAddressDeployer.js:106]: Same here.

File: test/ConstAddressDeployer.js [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+const deployContractConstant = async (deployer, wallet, contract, key, args = []) => {
+  const salt = getSaltFromKey(key);
+  const factory = new ContractFactory(
+      contract.abi,
+      contract.bytecode,
+  );
+  const bytecode = (await factory.getDeployTransaction(...args)).data;
+  await (await deployer.connect(wallet).deploy(bytecode, salt)).wait();
```
Comment (sammy1991106) [inline test/ConstAddressDeployer.js:30]: It's really hard to read when multiple `await`s exist in one statement.

**Review by sammy1991106 — CHANGES_REQUESTED** [review]

File: src/ConstAddressDeployer.sol (hunk 1 above)
Comment (Foivos) [reply src/ConstAddressDeployer.sol:10]: yes, the next argument (mload(bytecode)) is the length.

File: src/ConstAddressDeployer.sol (hunk 3 above)
Comment (Foivos) [reply src/ConstAddressDeployer.sol:7]: I think it needs to be memory.

**Review by re1ro — APPROVED** [review]

File: src/test/BurnableMintableCappedERC20Init.sol [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+pragma solidity 0.8.9;
+
+import { IAxelarGateway } from '../interfaces/IAxelarGateway.sol';
+import { IBurnableMintableCappedERC20 } from '../interfaces/IBurnableMintableCappedERC20.sol';
+
+import { MintableCappedERC20 } from '../MintableCappedERC20.sol';
+import { DepositHandler } from '../DepositHandler.sol';
+
+contract BurnableMintableCappedERC20Init is IBurnableMintableCappedERC20, MintableCappedERC20 {
```
Comment (re1ro) [inline src/test/BurnableMintableCappedERC20Init.sol:11]: Would it be possible just to extend `BurnableMintableCappedERC20` and define only constructor and init?

Comment (Foivos) [reply src/test/BurnableMintableCappedERC20Init.sol]: Done

Comment (Foivos) [reply src/test/BurnableMintableCappedERC20Init.sol]: done.

**Pull request closed without merging**